File: 프로그래머스/JavaScript/Lv. 1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
function solution(board, moves) {
  const basket = [];

  let cnt = 0;
  for (const move of moves) {
    let doll;
    for (let i = 0; i < board.length; i++) {
      if (board[i][move - 1] != 0) {
        doll = board[i][move - 1];
        board[i][move - 1] = 0;
        break;
      }
    }

    if (!doll) {
      continue;
    }

    let top = basket[basket.length - 1];
    if (top == doll && basket.length > 0) {
      basket.pop();
      cnt += 2;
    } else {
      basket.push(doll);
    }
  }

  return cnt;
}

File: 프로그래머스/Java/Lv. 1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.ArrayDeque;

class Solution {
	public int solution(int[][] board, int[] moves) {
		int answer = 0;

		ArrayDeque<Integer> basket = new ArrayDeque<Integer>();

		a: for (int i = 0; i < moves.length; i++) {
			int n = moves[i];

			for (int j = 0; j < board.length; j++) {
				int doll = board[j][n - 1];
				if (doll != 0) {
					if (!basket.isEmpty() && basket.peek() == doll) {
						basket.pop();
						answer += 2;
					} else {
						basket.push(doll);
					}

					board[j][n - 1] = 0;
					continue a;
				}
			}
		}

		return answer;
	}
}
